Protect against undefined content in Preview

ax-vasquez opened the pull request:
In Gatsby Cloud's Preview, all content changes are picked up by the extension. This means that newly-created posts will be sent to Gatsby _before_ the user has had a chance to add all required fields. Repositories for sites meant to be in use with Gatsby Preview should protect all references to required fields so that the site can handle missing elements gracefully.

## Changelog
* Protect against undefined content in Preview
* Update gatsby deps
* Fix npm vulnerabilities
* Exclude `dotenv` files from commits

axe312ger commented on gatsby-config.js:
```diff
@@ -23,7 +23,7 @@ const contentfulConfig = {
 if (process.env.CONTENTFUL_HOST) {
   contentfulConfig.host = process.env.CONTENTFUL_HOST;
-  contentfulConfig.accessToken = process.env.CONTENTFUL_PREVIEW_ACCESS_TOKEN;
+  contentfulConfig.accessToken = process.env.CONTENTFUL_DELIVERY_TOKEN;
```
You invert the logic users are used to:

If you have no host, we assume you want it from CDA (content delivery)

If you set a host, we assume you set it to CPA (content preview)

axe312ger replied:
Maybe change the way you set your environment variables?

ax-vasquez replied:
Thanks! I'll bring this up with the team and see what they suggest :+1:

ax-vasquez replied:
I've reverted this change for now, but am leaving this comment unresolved until I get some feedback from our product team regarding the environment variable change.

axe312ger replied:
👍

axe312ger commented on src/components/article-preview.js:
```diff
@@ -6,15 +6,27 @@ import styles from './article-preview.module.css'
+    { 
+      article.heroImage &&
+      article.heroImage.fluid &&
+      <Img alt="" fluid={article.heroImage.fluid} />
```
Gatsby supports https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining so lets use it :)

axe312ger commented on src/templates/blog-post.js:
```diff
@@ -16,27 +16,38 @@ class BlogPostTemplate extends React.Component {
           <Helmet title={`${post.title} | ${siteTitle}`} />
+          {
+          post.heroImage && 
+          post.heroImage.fluid && 
```
Same: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

axe312ger commented on src/templates/blog-post.js:
```diff
@@ -16,27 +16,38 @@ class BlogPostTemplate extends React.Component {
+              </p>
+            }
+            {
+              post.body &&
```
Same: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

axe312ger commented:
I'd like to have the userland version and this one aligned, at least on dependency/codestyle level.

Feel free to contact me in #gatsby-contentful in your Slack :)

ax-vasquez commented:
> I'd like to have the userland version and this one aligned, at least on dependency/codestyle level.
> 
> Feel free to contact me in #gatsby-contentful in your Slack :)
> 
> (Talking about [contentful-userland/gatsby-contentful-starter#90](https://github.com/contentful-userland/gatsby-contentful-starter/pull/90) which happened in parallel)

I _believe_ I addressed all of your feedback with the exception of the environment variable question (though, I did revert the change - I'm simply trying to make sure our understanding is in alignment with how you intend for us to use them :+1: )

With that said, I've since implemented:
* Optional chaining in spots that could break the Gatsby Preview when the field is undefined
* Set shared dependencies to the same versions as those used in the userland repo
* Add `prettier` as dev dependency and `.prettierrc` file with same settings as those in the userland repo
* Add `format` script to format code to that of the same settings used in the userland repo

ax-vasquez commented:
@axe312ger , could I get another review on this? :slightly_smiling_face: 

I _believe_ I addressed all of the feedback, but will certainly be happy to make changes as necessary :+1:

axe312ger commented on bin/setup.js:
```diff
@@ -1,11 +1,11 @@
-const spaceImport = require("contentful-import");
```
While I basically agree with single quote over double quote, please do not reformat these files within this PR.

axe312ger commented on bin/hello.js:
```diff
@@ -1,30 +1,30 @@
-const chalk = require("chalk");
```
please don't change this in here, make a separate PR

axe312ger commented on .prettierrc:
```diff
@@ -0,0 +1,6 @@
+{
```
its nice to add prettier, but thats unrelated to `protect against undefined`

Please remove these changes and make a separate PR

axe312ger commented on package.json:
```diff
@@ -45,6 +46,7 @@
     "netlify:login": "netlify login",
-    "netlify:deploy": "netlify deploy -d public -p"
+    "netlify:deploy": "netlify deploy -d public -p",
+    "format": "prettier --write 'src/**/*.js' 'bin/*.js'"
```
Same here. All upgrades and adding prettier is unrelated to your PR

ax-vasquez replied:
I know - I was confused as to why I needed to add styling stuff (per your comment above), but maybe I am misunderstanding? 

What did you mean when you said:
> I'd like to have the userland version and this one aligned, at least on dependency/codestyle level.

It sounded to me like you wanted me to implement the same code style settings as that userland repo PR, but I'm not really sure what the intent was :sweat_smile: I'll standby for feedback!

axe312ger reviewed (changes requested):
If you reduce your PR to the actually purpose of protecting from rendering undefined content, I happily give ✅ 

But please avoid adding extra unrelated stuff to a PR. Makes it hard to read, review and maintian.

axe312ger commented:
So if we remove the formatting of unrelated files and you settle the env variable question, lets merge this.


Feel free to contact me in slack via #gatsby-contentful channel

ax-vasquez commented:
> If you reduce your PR to the actually purpose of protecting from rendering undefined content, I happily give 
> 
> But please avoid adding extra unrelated stuff to a PR. Makes it hard to read, review and maintian.

Sorry, misunderstood what you meant when you asked for us to align this PR with the codestyle of the userland repo. I simply used the settings for that project. I'll go through this soon and get it fixed.

This is the comment I am referring to:
> "I'd like to have the userland version and this one aligned, at least on dependency/codestyle level."
